R/chat.R

```diff
@@ -49,50 +55,79 @@ NULL
+                            model = NA,
                             echo = FALSE) {
   check_string(system_prompt, allow_null = TRUE)
   check_string(base_url)
   check_string(api_key)
-  check_string(model)
+  check_string(model, allow_null = TRUE, allow_na = TRUE)
   check_bool(echo)
 
-  system_prompt <- system_prompt %||%
-    "You are a helpful assistant from New Zealand who is an experienced R programmer"
+  if (is.na(model)) {
+    model <- "gpt-4o-mini"
+  }
+
+  # TODO: Sanity check individual messages?
+  if (!is.null(messages)) {
+    if (!is.list(messages) ||
+        !(is.null(names(messages)) || all(names(messages) == ""))) {
+      stop_input_type(
+        messages,
+        "an unnamed list of messages",
+        allow_null = TRUE
+      )
+    }
+  }
+
+  if (!is.null(system_prompt)) {
+    if (!is.null(messages) && length(messages) > 0) {
+      if (messages[[1]][["role"]] != "system" ||
+          !identical(messages[[1]][["content"]], system_prompt)) {
+        stop("If both `system_prompt` and `messages` are provided, `messages` must start with the `system_prompt`.")
+      }
+    } else {
+      messages <- list(list(role = "system", content = system_prompt))
+    }
+  }
 
   ChatOpenAI$new(
     base_url = base_url,
     model = model,
+    messages = messages,
     api_key = api_key,
-    system_prompt = system_prompt,
     echo = echo
   )
 }
 
 #' @rdname new_chat_openai
 ChatOpenAI <- R6::R6Class("ChatOpenAI",
   public = list(
-    #' @param system_prompt A system prompt to set the behavior of the assistant.
     #' @param base_url The base URL to the endpoint; the default uses ChatGPT.
     #' @param api_key The API key to use for authentication. You generally should
     #'   not supply this directly, but instead set the `OPENAI_API_KEY` environment
     #'   variable.
     #' @param model The model to use for the chat; defaults to GPT-4o mini.
+    #' @param messages An unnamed list of messages to start the chat with (i.e.,
+    #'   continuing a previous conversation). If `NULL` or zero-length list, the
+    #'   conversation begins from scratch.
     #' @param echo If `TRUE`, the `chat()` method streams the response to stdout
     #'   (while also returning the final response). Note that this has no effect
     #'   on the `stream()`, `chat_async()`, and `stream_async()` methods.
-    initialize = function(base_url, model, api_key, system_prompt, echo = FALSE) {
+    initialize = function(base_url, model, messages, api_key, echo = FALSE) {
       private$base_url <- base_url
       private$model <- model
+      private$msgs <- messages %||% list()
       private$api_key <- api_key
       private$echo <- echo
+    },
 
-      private$add_message(list(
-        role = "system",
-        content = system_prompt
-      ))
+    #' @description The messages that have been sent and received so far
+    #'   (starting with the system prompt, if any).
+    messages = function() {
+      private$msgs
```

Comment:
Can you not keep this as `private$messges`?

Reply:
I need an official way to get them out, to serialize existing conversations. And I wanted it publicly read-only for now (although there are scenarios that will demand giving the ability to modify/transform the messages in some way, e.g. RAG, that we can hold off discussing til later)

Reply:
Sorry, I meant can't the private field name remain as `messages`?

Reply:
No, R6 explicitly checks that names are unique across private, public, and... a third one (protected, maybe?).

Reply:
Oooh, got it, thanks.